Sources/FoundationXML/XMLParser.swift

```diff
@@ -469,33 +461,35 @@ open class XMLParser : NSObject {
     open var externalEntityResolvingPolicy: ExternalEntityResolvingPolicy = .never
     
     open var allowedExternalEntityURLs: Set<URL>?
-    
-#if os(WASI)
-    private static var _currentParser: XMLParser?
-#endif
 
-    internal static func currentParser() -> XMLParser? {
-#if os(WASI)
-        return _currentParser
-#else
-        if let current = Thread.current.threadDictionary["__CurrentNSXMLParser"] {
-            return current as? XMLParser
-        } else {
-            return nil
+    /// The current parser is stored in a task local variable to allow for
+    /// concurrent parsing in different tasks with different parsers.
+    ///
+    /// Rationale for `@unchecked Sendable`:
+    /// While the ``XMLParser`` class itself is not `Sendable`, `TaskLocal`
+    /// requires the value type to be `Sendable`. The sendability requirement
+    /// of `TaskLocal` is only for the "default" value and values set with
+    /// `withValue` will not be shared between tasks.
+    /// So as long as 1. the default value is safe to be shared between tasks
+    /// and 2. the `Sendable` conformance of `_CurrentParser` is not used
+    /// outside of `TaskLocal`, it is safe to mark it as `@unchecked Sendable`.
+    private struct _CurrentParser: @unchecked Sendable {
+        let parser: XMLParser?
+
+        static var `default`: _CurrentParser {
+            return _CurrentParser(parser: nil)
         }
-#endif
+    }
+
+    @TaskLocal
```

Comment:
Are we always guaranteed to have a Task? For example, what happens if someone creates a thread (outside of swift concurrency) and calls `XMLParser.parse` - will it allow us to use Task-local storage?

Reply:
Yeah, TaskLocal falls back to TLS if the current execution is not bound to a Task. https://github.com/swiftlang/swift-evolution/blob/main/proposals/0311-task-locals.md#binding-task-local-values-from-synchronous-functions